### PR TITLE
Implement Function: Array.pick_random()

### PR DESCRIPTION
**Pull request by nonunknown**
Implements: https://github.com/godotengine/godot-proposals/issues/3454

This is partially implemented, since some users suggested adding for another classes, but I think each class should have their own PR, what you guys think?

**IMPORTANT:**
First merge #54940 before this one, to avoid https://github.com/godotengine/godot/issues/54939

test code:
```gdscript
var arr:Array = []

func _ready() -> void:
	arr.append_array(["test",1,2,3])
	print(arr)
	for i in range(10):
		print("Result: ", arr.pick_random())
```

output:

**First Run**:
```
[test, 1, 2, 3]
Result: 2
Result: 1
Result: 3
Result: 1
Result: 3
Result: 3
Result: 2
Result: 1
Result: 1
Result: 1

```

**Second Run**:
```
[test, 1, 2, 3]
Result: 1
Result: test
Result: 2
Result: 2
Result: 1
Result: 1
Result: 2
Result: 3
Result: 3
Result: 2

```

### CODE REVIEW

**Comment by sairam4123** [comment]
Would be interested in adding in probabilities too?
Like this?
`[1, 0, 2, 3].pick_random([0.25, 0.25, 0.25, 0.25])`
It's your choice to to decide whether the sum of all probabilities must be 1. It's how it works in Python.

**Comment by nonunknown** [comment]
> Would be interested in adding in probabilities too? Like this? `[1, 0, 2, 3].pick_random([0.25, 0.25, 0.25, 0.25])` It's your choice to to decide whether the sum of all probabilities must be 1. It's how it works in Python.

please create a proposal for that, every implementation depends on the users needs.

**Review by akien-mga — APPROVED** [review]
Approved in proposals review meeting.

File: core/variant/array.h [hunk 1]
```diff
@@ -80,6 +80,7 @@ class Array {
 
 	Variant front() const;
 	Variant back() const;
+	Variant pick_random() const;
```
Comment (akien-mga) [inline core/variant/array.h:83]: Trailing whitespace.

**Comment by akien-mga** [comment]
You need to update the class reference with `--doctool` and write documentation for the new method: https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html

Apparently there's also a style issue but not sure which, you should likely run clang-format.

**Comment by nonunknown** [comment]
@akien-mga just fixed everything!

File: doc/classes/Array.xml [hunk 2]
```diff
@@ -395,6 +395,17 @@
...
 		</method>
+		<method name="pick_random" qualifiers="const">
+			<return type="Variant" />
+			<description>
+				Returns a random value from the target array.
+				[codeblock]
+				var array:Array[int] = [1,2,3,4]
+				print(array.pick_random())
+				# output: 3
+				[/codeblock]
```
Comment (akien-mga) [inline doc/classes/Array.xml:406]:
```suggestion
				[codeblock]
				var array: Array[int] = [1, 2, 3, 4]
				print(array.pick_random())  # Prints either of the four numbers.
				[/codeblock]
```

File: core/variant/array.cpp [hunk 3]
```diff
@@ -31,6 +31,7 @@
 #include "array.h"
 
 #include "container_type_validate.h"
+#include "core/math/random_number_generator.h"
```
Comment (akien-mga) [inline core/variant/array.cpp:34]: You're not actually using this header here. `Math::rand()` is from `core/math/math_funcs.h`.

**Comment by akien-mga** [comment]
Please squash the commits when it's ready, we don't want 5 extra merge commits in the Git history.

**Comment by mhilbrunner** [comment]
Seems this includes a lot of other commits :) Needs a proper rebase.

**Comment by akien-mga** [comment]
@nonunknown has been struggling with rebasing on this one. Maybe a contributor can help salvage this?

**Pull request closed without merging**